usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -6446,34 +6888,51 @@ CK_RV ep11tok_unwrap_key(STDLL_TokData_t * tokdata, SESSION * session,
         rc = CKR_TEMPLATE_INCONSISTENT;
         goto error;
     }
-    switch (*(CK_OBJECT_CLASS *) cla_attr->pValue) {
-    case CKO_SECRET_KEY:
-        rc = check_key_attributes(tokdata,
-                                  *(CK_KEY_TYPE *) keytype_attr->pValue,
-                                  CKO_SECRET_KEY, attrs,
-                                  attrs_len, &new_attrs, &new_attrs_len, -1);
-        break;
-    case CKO_PUBLIC_KEY:
-        rc = check_key_attributes(tokdata,
-                                  *(CK_KEY_TYPE *) keytype_attr->pValue,
-                                  CKO_PUBLIC_KEY, attrs, attrs_len,
-                                  &new_attrs, &new_attrs_len, -1);
-        break;
-    case CKO_PRIVATE_KEY:
-        rc = check_key_attributes(tokdata,
-                                  *(CK_KEY_TYPE *) keytype_attr->pValue,
-                                  CKO_PRIVATE_KEY, attrs, attrs_len,
-                                  &new_attrs, &new_attrs_len, -1);
-        break;
-    default:
-        TRACE_ERROR("%s Missing CKA_CLASS type of wrapped key\n", __func__);
-        rc = CKR_TEMPLATE_INCOMPLETE;
-        goto error;
-    }
+    rc = check_ab_unwrap(tokdata, &verifyblob, &verifyblobsize, &vobj, kobj,
+                         attrs, attrs_len, mech, &isab);
     if (rc != CKR_OK) {
-        TRACE_ERROR("%s check key attributes failed: rc=0x%lx\n", __func__, rc);
+        TRACE_ERROR("check_ab_unwrap failed with rc=0x%08lx\n", rc);
         goto error;
     }
+    if (isab) {
+        rc = check_ab_kek_type(kobj->template, CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT);
+        if (rc != CKR_OK)
+            goto error;
+        /* AB unwrapping is taking care of all the boolean usage flags.
+         * No need to check attributes, just need to duplicate the array. */
+        if ((rc = dup_attribute_array(attrs,attrs_len,
+                                      &new_attrs, &new_attrs_len)))
+            goto error;
```

Comment:
For better reading I would add ( ...  != CKR_OK) or code it analog to rc = check_ab_kek_type above with a separate rc check in the next line

Reply:
Done.  That was actually a copy and paste from existing code.